src/lightning/data/streaming/shuffle.py

```diff
@@ -0,0 +1,167 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from functools import lru_cache
+from typing import Any, List
+
+import numpy as np
+
+from lightning.data.datasets.env import _DistributedEnv
+from lightning.data.streaming import Cache
+
+
+class Shuffle(ABC):
+    """Shuffle describe how to distribute chunked datasets across processes and workers."""
+
+    def __init__(self, cache: Cache, seed: int):
+        self.cache = cache
+        self.seed = seed
+        self.random_state = None
+
+    @abstractmethod
+    def get_len(self, distributed_env: _DistributedEnv, current_epoch: int) -> int:
+        pass
+
+    @abstractmethod
+    def get_chunks_and_intervals_per_process(self, distributed_env: _DistributedEnv, current_epoch: int) -> Any:
+        pass
+
+    @abstractmethod
+    def __call__(self, array: np.ndarray) -> List[int]:
+        pass
+
+
+class NoShuffle(Shuffle):
+    """NoShuffle doesn't shuffle the items and ensure all the processes receive the same number of items."""
+
+    @lru_cache(maxsize=10)
+    def get_len(self, distributed_env: _DistributedEnv, current_epoch: int) -> int:
+        _, intervals_per_process = self.get_chunks_and_intervals_per_process(distributed_env, current_epoch)
+        min_items_per_process = min(
+            [sum([(interval[-1] - interval[0]) for interval in intervals]) for intervals in intervals_per_process]
+        )
+        return min_items_per_process
+
+    @lru_cache(maxsize=10)
+    def get_chunks_and_intervals_per_process(self, distributed_env: _DistributedEnv, current_epoch: int) -> Any:
+        self.random_state = np.random.RandomState(seed=self.seed + current_epoch)  # type: ignore
+        chunk_intervals = self.cache.get_chunk_intervals()
+        indexes = list(range(len(chunk_intervals)))
+        shuffled_chunk_intervals = np.asarray(chunk_intervals)[indexes]
+
+        chunks_per_process: List[List[int]] = [[] for _ in range(distributed_env.world_size)]
+        intervals_per_process: List[List[List[int]]] = [[] for _ in range(distributed_env.world_size)]
+        for index, (chunk_index, chunk_interval) in enumerate(zip(indexes, shuffled_chunk_intervals)):
+            replica_index = index % distributed_env.world_size
+            chunks_per_process[replica_index].append(chunk_index)
+            intervals_per_process[replica_index].append(chunk_interval)
+
+        return chunks_per_process, intervals_per_process
+
+    def __call__(self, array: np.ndarray) -> List[int]:
+        return array.tolist()
+
+
+class MinShuffle(Shuffle):
```

Comment:
I don't think the name is explanatory - `Min` with respect to what? It makes it hard for users to decide what to do.
Also the docstring is exactly the same as `AcrossChunkShuffle`.

Honestly I would just go for no shuffle or what we call "min" shuffle today.
If I understand it well, the across chunks would be quite inefficient and it's almost an antipattern for streaming IMO.

Shuffling over one or just a few chunks is really what we should push users to choose.

Reply:
We need to support both. The default is the `MinShuffle` but we can't not offer a shuffling solution that doesn't lose that many samples.

Reply:
I get it now, yes makes sense.
Let’s maybe just find a better name for Min.
Also AcrossChunk is misleading IMO